man/coq_vo2xml.1
.TH COQ 1 "April 25, 2001"

.SH NAME
coq_vo2xml \- Exports the content of a Coq module to XML


.SH SYNOPSIS
.B coq_vo2xml
[
.B general \ Coq \ options
]
[
.B -image
.I f
]
[
.B -t
]
[
.B -xml-library-root
.I d
]
.IR module " .\|.\|."


.SH DESCRIPTION

.B coq_vo2xml
is a tool to export the content of a Coq module to XML.
Because it executes coqtop(1), it recognizes a superset of its options.
ones:
.PP
.B -image
.IR f
overrides the coqtop(1) executable with
.IR f "."
.PP
.B -t
does not remove the temporary file fed to coqtop(1).
.PP
.B -xml-library-root
.IR d
sets the path to the root of the XML library to
.IR d ","
overriding the value of $XML_ROOT_LIBRARY. The root of the XML library
must be specified, either using the flag or setting the variable.
.PP
.IR module \&
is the name of the module to export. The file module.vo will be searched in the
Coq path by coqtop(1).
The compiler produces a whole tree of XML files and directories rooted in the
specified XML root library.

.SH OPTIONS

.TP
.BI \-h
Will give you a description of the whole list of options of coq_vo2xml,
including the ones of coqtop(1).

.SH SEE ALSO

.BR coqtop (1),
.BR coq_makefile (1),
.br
.I
The Coq Reference Manual.
.I
The Coq web site: http://coq.inria.fr


CHANGES
Changes from V8.4
=================

Vernacular commands

- The command "Record foo ..." does not generate induction principles
  (foo_rect, foo_rec, foo_ind) anymore by default (feature wish #2693).
  A flag "Set/Unset Record Elimination Schemes" allows to change this.
  The tactic "induction" on a record is now actually doing "destruct".
- The "Open Scope" command can now be given also a delimiter (e.g. Z).

Specification Language

- The syntax "x -> y" is now declared at level 99. In particular, it has
  now a lower priority than "<->" : "A -> B <-> C" is now "A -> (B <-> C)".

Tactics

- Tactics btauto, a reflexive boolean tautology solver.
- Tactic "tauto" was exceptionally able to destruct other connectives
  than the binary connectives "and", "or", "prod", "sum", "iff".  This
  non-uniform behavior has been fixed (bug #2680) and tauto is
  slightly weaker. On the opposite side, new tactic "dtauto" is able
  to destruct any record-like inductive types, superseding the old
  version of "tauto".
- Similarly, "intuition" has been made more uniform and, where it now
  fails, "dintuition" can be used.

Program

- "Solve Obligations using" changed to "Solve Obligations with",
  consistent with "Proof with".
- Program Lemma, Definition now respect automatic introduction.

Notations

- "Bind Scope" can no longer bind "Funclass" and "Sortclass".
- A notation can be given a (compat "8.x") annotation, making
  it behave like a (only parsing), but flags may active warning
  or error when this notation is used.

Internal Infrastructure

- Many reorganisations in the ocaml source files. For instance,
  many internal a.s.t. of Coq are now placed in mli files in
  a new directory intf/, for instance constrexpr.mli or glob_term.mli.
  More details in dev/doc/changes.

Changes from V8.4beta2 to V8.4
==============================

Vernacular commands

- The "Reset" command is now supported again in files given to coqc or Load.
- "Show Script" now indents again the displayed scripts. It can also work
  correctly across Load'ed files if the option "Unset Atomic Load" is used.
- "Open Scope" can now be given the delimiter (e.g. Z) instead of the full
  scope name (e.g. Z_scope).

Notations

- Most compatibility notations of the standard library are now tagged as
  (compat xyz), where xyz is a former Coq version, for instance "8.3".
  These notations behave as (only parsing) notations, except that they may
  triggers warnings (or errors) when used while Coq is not in a corresponding
  -compat mode.
- To activate these compatibility warnings, use "Set Verbose Compat Notations"
  or the command-line flag -verbose-compat-notations.
- For a strict mode without these compatibility notations, use
  "Unset Compat Notations" or the command-line flag -no-compat-notations.

Tactics

- An annotation "eqn:H" or "eqn:?" can be added to a "destruct"
  or "induction" to make it generate equations in the spirit of "case_eq".
  The former syntax "_eqn" is discontinued.
- The name of the hypothesis introduced by tactic "remember" can be
  set via the new syntax "remember t as x eqn:H" (wish #2489).

Libraries

- Reals: changed definition of PI, no more axiom about sin(PI/2).
- SetoidPermutation: a notion of permutation for lists modulo a setoid equality.
- BigN: fixed the ocaml code doing the parsing/printing of big numbers.

Changes from V8.4beta to V8.4beta2
==================================

Vernacular commands

- Back and BackTo are now handling the proof states. They may
  perform some extra steps of backtrack to avoid states where
  the proof state is unavailable (typically a closed proof).
- The commands Suspend and Resume have been removed.
- A basic Show Script has been reintroduced (no indentation).
- New command "Set Parsing Explicit" for deactivating parsing (and printing)
  of implicit arguments (useful for teaching).
- New command "Grab Existential Variables" to transform the unresolved evars at
  the end of a proof into goals.

Tactics

- Still no general "info" tactical, but new specific tactics
  info_auto, info_eauto, info_trivial which provides information
  on the proofs found by auto/eauto/trivial. Display of these
  details could also be activated by Set Info Auto/Eauto/Trivial.
- Details on everything tried by auto/eauto/trivial during
  a proof search could be obtained by "debug auto", "debug eauto",
  "debug trivial" or by a global "Set Debug Auto/Eauto/Trivial".
- New command "r string" that interprets "idtac string" as a breakpoint
  and jumps to its next use in Ltac debugger.
- Tactics from the Dp plugin (simplify, ergo, yices, cvc3, z3, cvcl,
  harvey, zenon, gwhy) have been removed, since Why2 has not been
  maintained for the last few years. The Why3 plugin should be a suitable
  replacement in most cases.

Libraries

- MSetRBT: a new implementation of MSets via Red-Black trees (initial
  contribution by Andrew Appel).
- MSetAVL: for maximal sharing with the new MSetRBT, the argument order
  of Node has changed (this should be transparent to regular MSets users). 

Module System

- The names of modules (and module types) are now in a fully separated
  namespace from ordinary definitions: "Definition E:=0. Module E. End E."
  is now accepted.

CoqIDE

- Coqide now supports the Restart command, and Undo (with a warning).
  Better support for Abort.

Changes from V8.3 to V8.4beta
=============================

Logic

- Standard eta-conversion now supported (dependent product only) (DOC TO DO).
- Guard condition improvement: subterm property is propagated through beta-redex
  blocked by pattern-matching, as in "(match v with C .. => fun x => u end) x";
  this allows for instance to use "rewrite ... in ..." without breaking
  the guard condition.

Specification language and notations

- Maximal implicit arguments can now be set locally by { }. The registration
  traverses fixpoints and lambdas. Because there is conversion in types,

tools/coq_vo2xml.ml
(* environment *)

let environment = Unix.environment ()

let bindir = ref Coq_config.bindir
let binary = ref ("coqtop." ^ Coq_config.best)
let image = ref ""
let xml_library_root = ref (
 try
  Sys.getenv "XML_LIBRARY_ROOT"
 with Not_found -> ""
)

(* the $COQBIN environment variable has priority over the Coq_config value *)
let _ = 
  try 
    let c = Sys.getenv "COQBIN" in
    if c <> "" then bindir := c
  with Not_found -> ()

(* coq_vo2xml options *)

let keep = ref false

(* Verifies that a string do not contains others caracters than letters, 
   digits, or `_` *)

let check_module_name s = 
  let err () = 
    output_string stderr
      "Modules names must only contain letters, digits, or underscores\n"; 
    output_string stderr
      "and must begin with a letter\n";
    exit 1 
  in
  match String.get s 0 with 
    | 'a' .. 'z' | 'A' .. 'Z' -> 
	for i = 1 to (String.length s)-1 do
	  match String.get s i with 
	    | 'a' .. 'z' | 'A' .. 'Z' | '0' .. '9' | '_'  -> ()
	    | _ -> err ()
	done
    | _ -> err ()

 (* compilation of a file [file] with command [command] and args [args] *)

let compile command args file =
  let dirname = Filename.dirname file in
  let basename = Filename.basename file in
  let modulename =
    if Filename.check_suffix basename ".vo" then
      Filename.chop_suffix basename ".vo"
    else
      basename 
  in
  check_module_name modulename;
  let tmpfile = Filename.temp_file "coq_vo2xml" ".v" in
  let args' = 
    command :: "-batch" :: "-silent" :: args 
    @ ["-load-vernac-source"; tmpfile] in
  let devnull = 
    if Sys.os_type = "Unix" then
      Unix.openfile "/dev/null" [] 0o777 
    else 
      Unix.stdin
  in 
  let oc = open_out tmpfile in
  Printf.fprintf oc "Require Xml.\n" ;
  Printf.fprintf oc "Require %s.\n" modulename;
  Printf.fprintf oc "Print XML Module Disk \"%s\" %s.\n" !xml_library_root
   modulename;
  flush oc;
  close_out oc;
  try
    let pid =
      Unix.create_process_env command
        (Array.of_list args') environment devnull Unix.stdout Unix.stderr in
    let status = Unix.waitpid [] pid in
    if not !keep then Sys.remove tmpfile ;
    match status with
      | _, Unix.WEXITED 0 -> ()
      | _, Unix.WEXITED 127 -> 
	  Printf.printf "Cannot execute %s\n" command;
	  exit 1
      | _, Unix.WEXITED c -> exit c
      | _                 -> exit 1
  with _ -> 
    if not !keep then Sys.remove tmpfile; exit 1

(* parsing of the command line
 *
 * special treatment for -bindir and -i.
 * other options are passed to coqtop *)

let usage () =
  Usage.print_usage
   "Usage: coq_vo2xml <options> <Coq options> module...\n
options are:
  -xml-library-root d   specify the path to the root of the XML library
                        (overrides $XML_LIBRARY_ROOT)
  -image f              specify an alternative executable for Coq
  -t                    keep temporary files\n\n" ;
  flush stderr ;
  exit 1

let parse_args () =
  let rec parse (cfiles,args) = function
    | [] -> 
	List.rev cfiles, List.rev args
    | "-xml-library-root" :: v :: rem ->
        xml_library_root := v ; parse (cfiles,args) rem
    | "-t" :: rem -> 
	keep := true ; parse (cfiles,args) rem
    | "-boot" :: rem ->
	bindir:= Filename.concat Coq_config.coqtop "bin";
	parse (cfiles, "-boot"::args) rem
    | "-bindir" :: d :: rem ->
	bindir := d ; parse (cfiles,args) rem
    | "-bindir" :: []       ->
	usage ()
    | "-byte" :: rem ->
	binary := "coqtop.byte"; parse (cfiles,args) rem
    | "-opt" :: rem ->
	binary := "coqtop.opt"; parse (cfiles,args) rem
    | "-image" :: f :: rem ->
	image := f; parse (cfiles,args) rem
    | "-image" :: [] ->
	usage ()
    | ("-?"|"-h"|"-H"|"-help"|"--help") :: _ -> usage ()
    | ("-libdir"|"-outputstate"|"-I"|"-include"
      |"-inputstate"|"-is"|"-load-vernac-source"|"-load-vernac-object"
      |"-load-ml-source"|"-require"|"-load-ml-object"|"-user"
      |"-init-file" as o) :: rem ->
	begin
	  match rem with
	    | s :: rem' -> parse (cfiles,s::o::args) rem'
	    | []        -> usage ()
	end
    | "-R" as o :: s :: t :: rem -> parse (cfiles,t::s::o::args) rem
    | ("-notactics"|"-debug"|"-db"|"-debugger"|"-nolib"|"-batch"|"-nois"
      |"-q"|"-full"|"-profile"|"-just-parsing"|"-echo" |"-unsafe"|"-quiet"
      |"-silent"|"-m" as o) :: rem ->
	parse (cfiles,o::args) rem
    | ("-v"|"--version") :: _ ->
        Usage.version ()
    | "-where" :: _ -> 
	print_endline Coq_config.coqlib; exit 0
    | f :: rem -> parse (f::cfiles,args) rem
  in
  parse ([],[]) (List.tl (Array.to_list Sys.argv))

(* main: we parse the command line, define the command to compile files
 * and then call the compilation on each file *)

let main () =
  let cfiles, args = parse_args () in
  if cfiles = [] then begin
    prerr_endline "coq_vo2xml: too few arguments" ;
    usage ()
  end;
  let coqtopname = 
    if !image <> "" then !image else Filename.concat !bindir (!binary ^ Coq_config.exec_extension)
  in
  if !xml_library_root = "" then
   begin
    prerr_endline "coq_vo2xml: you must either set $XML_LIBRARY_ROOT or use -xml-library-root";
    usage ()
   end
  else
   List.iter (compile coqtopname args) cfiles ;
   prerr_endline
    ("\nWARNING: all the URIs in the generated XML files are broken." ^
     "\n         See the README in the XML contrib to learn how to fix them.\n")
    
let _ = Printexc.print main (); exit 0
